### PR TITLE
Sort by timestamp

### PR DESCRIPTION
**Pull request by smellthemoon**
Signed-off-by: lixinguo <xinguo.li@zilliz.com>
#19853

### CODE REVIEW

File: internal/querynode/result_sorter.go [hunk 1]
```diff
@@ -1,41 +1,48 @@
...
-
-func (s *byPK) Swap(i, j int) {
-	s.r.Offset[i], s.r.Offset[j] = s.r.Offset[j], s.r.Offset[i]
-
-	typeutil.SwapPK(s.r.GetIds(), i, j)
-
-	for _, field := range s.r.GetFieldsData() {
-		swapFieldData(field, i, j)
+func GetTimeStampFieldsData(s *segcorepb.RetrieveResults) []Timestamp {
+	if s.FieldsData != nil {
```
Comment (jiaoew1991) [inline internal/querynode/result_sorter.go:32]: reverse the if and return first

File: internal/querynode/result_sorter.go [hunk 2]
```diff
@@ -85,3 +92,25 @@ func swapFieldData(field *schemapb.FieldData, i int, j int) {
 		panic(errMsg)
 	}
 }
+
+func (s *byTS) Len() int {
+	if s.r == nil {
```
Comment (jiaoew1991) [inline internal/querynode/result_sorter.go:97]: why not return `len(s.ts)` directly? when s.r is nil `len(s.ts)` should be 0?

**Review by XuanYang-cn — CHANGES_REQUESTED** [review]
recommend keeping both sortByPK and sortByTS

File: internal/proxy/task_search.go [hunk 3]
```diff
@@ -328,6 +327,8 @@ func (t *searchTask) PreExecute(ctx context.Context) error {
 			return err
 		}
 
+		t.request.OutputFields = append(t.request.OutputFields, common.TimeStampFieldName)
+		outputFieldIDs = append(outputFieldIDs, common.TimeStampField)
 		t.SearchRequest.OutputFieldsId = outputFieldIDs
 		plan.OutputFieldIds = outputFieldIDs
```
Comment (XuanYang-cn) [inline internal/proxy/task_search.go:334]: Changes in wrong function, change task_query.go instead

File: internal/proxy/task_search.go [hunk 4]
```diff
@@ -328,6 +327,8 @@ func (t *searchTask) PreExecute(ctx context.Context) error {
 			return err
 		}
 
+		t.request.OutputFields = append(t.request.OutputFields, common.TimeStampFieldName)
+		outputFieldIDs = append(outputFieldIDs, common.TimeStampField)
```
Comment (jiaoew1991) [inline internal/proxy/task_search.go:331]: it means the output fields must contain the TS field. Is it right?

File: internal/proxy/task_search.go [hunk 5]
```diff
@@ -328,6 +327,8 @@ func (t *searchTask) PreExecute(ctx context.Context) error {
...
 		t.SearchRequest.OutputFieldsId = outputFieldIDs
```
Comment (XuanYang-cn) [inline internal/proxy/task_search.go:332]: The user doesn't need the timestamp filed as output field, please remove this field at proxy

**Review by XuanYang-cn — CHANGES_REQUESTED** [review]

File: internal/querynode/segment.go [hunk 6]
```diff
@@ -411,7 +411,8 @@ func (s *Segment) retrieve(plan *RetrievePlan) (*segcorepb.RetrieveResults, erro
 		return nil, err
 	}
 
-	sort.Sort(&byPK{result})
+	timeStamp := GetTimeStampFieldsData(result)
+	sort.Sort(&byTS{result, timeStamp})
```
Comment (XuanYang-cn) [inline internal/querynode/segment.go:415]: Please don't change the result order unless you change all the reducing logic along with it. Or the result after this PR is meaningless.

**Review by XuanYang-cn — CHANGES_REQUESTED** [review]

File: internal/proxy/task_search.go (hunk 3 above)
Comment (smellthemoon) [reply internal/proxy/task_search.go:334]: ok

File: internal/proxy/task_search.go (hunk 5 above)
Comment (smellthemoon) [reply internal/proxy/task_search.go:332]: ok,I will delete it at proxy

**Pull request closed without merging**